Modules/_csv.c

```diff
@@ -326,6 +340,22 @@ static char *dialect_kws[] = {
     NULL
 };
 
+static _csvstate *
+_csv_state_from_type(PyTypeObject *type, const char *name)
+{
+    PyObject *module = _PyType_GetModuleByDef(type, &_csvmodule);
+    if (module == NULL) {
+        return NULL;
+    }
+    _csvstate *module_state = PyModule_GetState(module);
+    if (module_state == NULL) {
+        PyErr_Format(PyExc_SystemError,
+                     "%s: No _csv module state found", name);
```

Comment:
I would expect that _csv_state_from_type() cannot fail (return NULL) and so use assertions rather than runtime checks (if).

If you consider that it *can* fail, you should check if _csv_state_from_type() fails at each _csv_state_from_type() call, no? For example, dialect_new() uses module_state without checking if it is NULL.

I didn't check if _csv_state_from_type() can be called on a type which is a subtype of the *type* parameter.

Well, in case of doubt, just check for _csv_state_from_type() failure ;-)

Reply:
> I didn't check if _csv_state_from_type() can be called on a type which is a subtype of the type parameter.

It can; that's why this uses `_PyType_GetModuleByDef`.
I'll add the type check.